### PR TITLE
Add aws client

### PR DESCRIPTION
**Pull request by jiayiwang7**
*Issue #, if available:*

#1573  

*Description of changes:*

AWS Client for snow preflight validation.

*Testing (if applicable):*

unit testing

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pkg/aws/client.go
Comment (mitalipaygude) [inline pkg/aws/client.go:48]: nit: can we just return `nil, fmt.Errorf("setting aws config: %v", err)`?

File: pkg/aws/creds.go
Comment (mitalipaygude) [inline pkg/aws/creds.go:77]: nit: Is there a reason why we are using `errors.New()` as well as `fmt.Errorf` in the same pkg? Should we be consistent with creating those?

File: pkg/aws/ec2.go
Comment (chrisdoherty4) [inline pkg/aws/ec2.go:36]: This error results in an arbitrary hash that the user must assume represents an image ID. Easily cleared up with a key.

```suggestion
	return false, fmt.Errorf("aws describe image [imageID=%s]: %v", imageID, err)
```

We should also agree on whether errors use square or rounded parenthesis.

File: pkg/aws/ec2.go
Comment (chrisdoherty4) [inline pkg/aws/ec2.go:39]: I'm not sure a key name being on the target means the key pair exists/is valid. Could you elaborate on expectations here? My train of thought is that key names could be different between my machine and the host machine but the key pairs are still valid. 

Additionally, why does a client perform this check? I would think the client merely provides behavior for interacting with the target but the actual "does this exist on the target" logic would exist elsewhere. Concretely, a client could provide a `RetrieveKeysByName()` method and some other (probably a validating construct) uses that to retrieve the keys and then performs the necessary checks to ensure the key pair is valid (i.e. the business logic). Why is this important? Your validator is now isolatable because you can use an interface to mock up the `RetrieveKeysByName()` and test the business logic very easily.

File: pkg/aws/snow.go
Comment (chrisdoherty4) [inline pkg/aws/snow.go:47]:
```suggestion
	return aws.EndpointResolverWithOptionsFunc(func(service, region string, options ...interface{}) (aws.Endpoint, error) {
		endpoint := snowEndpoint(deviceIP)
		if service == endpoint.ServiceID {
```

File: pkg/aws/ec2.go
Comment (chrisdoherty4) [inline pkg/aws/ec2.go:23]: Using `a` as the receiver parameter which doesn't seem to be relatable to `*client` in any capacity.

File: pkg/aws/snow.go
Comment (chrisdoherty4) [inline pkg/aws/snow.go:32]: This is OK but there is a nuance in behavior with respect to what a maintainer intuitively expects (probably).

Intuitively I feel a maintainer, given the `WithCustomCABundleFile` doesn't return an error, would assume the file is read _when the option is applied_. That gives cause for me to, hypothetically, write a file _after_ calling `WithCustomCABundleFile` but _before_ the option is applied. While that may seem silly, its quite legitimate when reflecting on the package API and would result in odd behavior because the option would always error.

Below is a suggestion to defer reading the file and still apply the option. It just constructs the actual option in-line when its needed. I also took the liberty to swap out reading the entire file into memory with a buffered io stream to be a little more efficient (probably negligable given the pem won't be huge but its good practice to prefer streams nonetheless; the buffered reader is likely unnecessary too but doesn't hurt). Finally, I changed the error to include the root error as `%w` meaning consumers can get at the root cause if they desire (its a standard library error so doesn't hurt making it part of the pkg api).

```suggestion
	return func(opts *config.LoadOptions) error {
		caPEM, err := os.Open(certsFile)
		if err != nil {
			return fmt.Errorf("reading aws certificates file: %w", err)
		}
		
		customBundleOpt := config.WithCustomCABundle(bufio.NewReader(caPEM))
		if err := customBundleOpt(opts); err != nil {
			return err
		}
		
		return 
	}
```

File: pkg/aws/snow.go
Comment (chrisdoherty4) [inline pkg/aws/snow.go:24]: For consistency on this package API, we probably want to return an `AwsConfigOpt`. The compiler shouldn't complain because its an alias.

**Comment by chrisdoherty4** [comment]
All exported package names should have good documentation and documenting non trivial unexported names is strongly encouraged.

File: pkg/aws/snow.go
Comment (g-gaston) [inline pkg/aws/snow.go:35]: maybe `WithSnowEndpoint` or `WithSnowEndpointAccess`?

File: pkg/aws/snow_test.go
Comment (g-gaston) [inline pkg/aws/snow_test.go:1]: can we test this from `aws_test`? or do the tests depend on something unexported?

File: pkg/dependencies/testdata/cluster_snow.yaml
Comment (g-gaston) [inline pkg/dependencies/testdata/cluster_snow.yaml:1]: Is this file being used?

File: pkg/providers/factory/providerfactory.go
Comment (g-gaston) [inline pkg/providers/factory/providerfactory.go:28]: Is this being used?

File: pkg/aws/client.go
Comment (jiayiwang7) [reply pkg/aws/client.go:48]: no it's not a pointer

File: pkg/aws/ec2.go
Comment (g-gaston) [reply pkg/aws/ec2.go:39]: If we go this route, I would suggest we make sure the client doesn't return any aws sdk specific type.

Comment (jiayiwang7) [reply pkg/aws/ec2.go:39]: this is amazon ec2 key pairs: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-key-pairs.html

File: pkg/aws/client.go
Comment (chrisdoherty4) [inline pkg/aws/client.go:13]: Is there a need to export this?

File: pkg/aws/ec2_test.go
Comment (g-gaston) [inline pkg/aws/ec2_test.go:33]: nit: we might want to create a constructor to avoid having to export this

**Review by g-gaston — APPROVED** [review]
/lgtm
/hold

File: pkg/aws/ec2.go
Comment (chrisdoherty4) [reply pkg/aws/ec2.go]: #### EC2 key pairs

I read the function name as checking if a key pair exists hence felt the behavior was wrong. Looking at the replies and docs you linked, I think the method name is wrong. We're only trying to provide a guarentee that the key name they supplied exists on the EC2 instance so the method should reflect that: `KeyNameExists()`. Does that resonate?

Off the back of that, all the methods seem to be EC2 specific which makes me think the struct name isn't specific enough - could we rename it to `EC2Client`?

--- 

Re behavior splitting among constructs: I empathize with your sentiment - 👍🏻. I don't know if its the right thing to do because "[doesn't contain] much validation logic" > 0 which means this construct both interfaces with EC2 APIs and validates stuff hence could be violating the SRP. On the flip side, these could just be framed as convenience methods on the EC2 API so I commit!

File: pkg/aws/creds.go
Comment (chrisdoherty4) [reply pkg/aws/creds.go:77]: Take a look at the code for [`fmt.Errorf`](https://cs.opensource.google/go/go/+/go1.18:src/fmt/errors.go;l=17) - invoking that unnecessarily is, less than optimal relative to the simple [`errors.New`](https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/errors/errors.go;l=58) 😆.

File: pkg/aws/ec2.go
Comment (jiayiwang7) [reply pkg/aws/ec2.go]: 1. I feel we should stick to KeyPairExists. We check the aws ec2 key pair existence by the key name.
2. For now it is ec2 specific but I want this struct to be general enough to handle other aws services.

**Pull request merged**
